Update to Cabal 3.2.0.0

<!--
  Have you read our Code of Conduct?
  By filing an issue, you are expected to comply with it, including treating everyone with respect:
  https://github.com/FreeProving/guidelines/blob/main/CODE_OF_CONDUCT.md
-->

## Description

<!-- A brief description of the feature. -->
We should upgrade to Cabal 3.2.0.0.

## Motivation

<!--
  Why are we doing this?
  What use cases does it support?
  What is the expected outcome?
-->
In addition to the reasons stated in <https://github.com/FreeProving/free-compiler/issues/63> we have to update to Cabcal 3.2 due to a bug in Cabal 2.4. 

As I've found out in #28 Haddock is incompatible with Compiler plugins. If we still want to use the Polysemy plugin, it has to be disabled in the `./tool/make-docs.sh` script. For this purpose, Cabal build flags have to be set in the script and `ghc-options` in the Cabal file have to be conditional. In #28 the pipeline passes but the fix breaks the `./tool/run.sh` script added in #27. The `cabal new-run` command apparently ignores conditional blocks in Cabal 2.4. In Cabal 3.2 this bug seems to be fixed.

## Alternatives

<!--
  What alternative solutions have you considered?
  What are their drawbacks in comparison to your proposal?
-->
We could disable the Polysemy plugin but would have to use explicit `TypeApplications` in some places.

## Additional Context

<!-- Add any other context about the feature request here. -->
<details>
<summary>
<code>./tool/make.sh</code> error with <code>-fplugin=Polysemy.Plugin</code>
</summary>

```
<command line>: Could not find module ‘Polysemy.Plugin’
Use -v to see a list of the files searched for.
```
</details>

<details>
<summary>
<code>./tool/run.sh</code> error with conditional <code>-fplugin=Polysemy.Plugin</code>
</summary>

```
src/lib/HST/Effect/Cancel.hs:55:13: error:
    • Ambiguous use of effect 'Polysemy.Error.Error'
      Possible fix:
        add (Member (Polysemy.Error.Error ()) (Polysemy.Error.Error a0
                                                 : r)) to the context of 
          the type signature
      If you already have the constraint you want, instead
        add a type application to specify
          '()' directly, or activate polysemy-plugin which
            can usually infer the type correctly.
    • In the expression: throw ()
      In a case alternative: Cancel -> throw ()
      In the first argument of ‘reinterpret’, namely
        ‘\case Cancel -> throw ()’
   |
55 |   Cancel -> throw ()
   |             ^^^^^^^^
cabal: Failed to build haskell-src-transformations-0.1.1.0 (which is required
by exe:haskell-src-transformations from haskell-src-transformations-0.1.1.0).
```
</details>

--- a/.github/workflows/ci-pipeline.yml
+++ b/.github/workflows/ci-pipeline.yml
@@ -81,7 +81,7 @@ jobs:
       uses: actions/setup-haskell@v1.1
       with:
         ghc-version: '8.6.5'
-        cabal-version: '2.4'
+        cabal-version: '3.2'
     - name: Update Cabal Package List
       id: cabal-update
       run: cabal new-update
@@ -167,7 +167,7 @@ jobs:
       uses: actions/setup-haskell@v1.1
       with:
         ghc-version: '8.6.5'
-        cabal-version: '2.4'
+        cabal-version: '3.2'
     - name: Update Cabal Package List
       id: cabal-update
       run: cabal new-update
@@ -224,7 +224,7 @@ jobs:
       uses: actions/setup-haskell@v1.1
       with:
         ghc-version: '8.6.5'
-        cabal-version: '2.4'
+        cabal-version: '3.2'
     - name: Update Cabal Package List
       id: cabal-update
       run: cabal new-update

--- a/README.md
+++ b/README.md
@@ -53,7 +53,7 @@ To build it, the GHC and Cabal are required.
 The tool has been tested with the following software versions.
 
  - [GHC][software/ghc], version  8.6.5
- - [Cabal][software/cabal], version 2.4.1.0
+ - [Cabal][software/cabal], version 3.2.0.0
 
 ### Executable Installation
 

--- a/haskell-src-transformations.cabal
+++ b/haskell-src-transformations.cabal
@@ -44,15 +44,22 @@ common deps
   --
   -- Polysemy needs a lot of language extensions. The language extensions
   -- which are required in every computation that uses Polysemy effects
-  -- are enabled globally. Language extensions such as `TemplateHaskell`
+  -- are enabled by default. Language extensions such as `TemplateHaskell`
   -- which are needed for the definition of custom effects only are enabled
-  -- for individual files.
+  -- for individual files and are listed as `other-extensions`.
   default-language:    Haskell2010
-  extensions:          DataKinds
+  default-extensions:  DataKinds
                      , FlexibleContexts
                      , GADTs
                      , PolyKinds
                      , TypeOperators
+  other-extensions:    BlockArguments
+                     , FlexibleContexts
+                     , LambdaCase
+                     , ScopedTypeVariables
+                     , StandaloneDeriving
+                     , TemplateHaskell
+                     , TypeFamilies
   ghc-options:         -Wall
 
   -- The Polysemy plugin is currently not supported by Haddock. Thus, the

--- a/tool/test.sh
+++ b/tool/test.sh
@@ -22,9 +22,7 @@ cd "$root_dir"
 # add the `--rerun` option easily.
 failure_report_file=".hspec-failure-report"
 
-# Run tests with cabal.
-# We use `new-run` instead of `new-test`, because Cabal 2.4 does not
-# support `--test-options` to be passed to the test.
+# Run tests with Cabal.
 cabal new-run haskell-src-transformations-unit-tests \
   --ghc-option -Wwarn --                             \
   --failure-report="$failure_report_file"            \

